[LVI] Infer ranges from mul nuw square conditions

ken-matsui opened the pull request:
A non-poison comparison involving `mul nuw X, X` implies that the
multiplication does not overflow.  This bounds X by:

    X <= floor(sqrt(2^bitwidth(X) - 1)) (e.g., i16: X <= 255)

An unsigned constant comparison can tighten the bound, e.g.,
`X * X <= 120` implies `X <= 10`.

Fixes https://github.com/llvm/llvm-project/issues/122412

ken-matsui committed:
Add baseline tests for upcoming patch

nikic reviewed (commented):
I'd rather handle this in ValueTracking or LazyValueInfo, especially for this case where the reasoning is in terms of the bit width only. We haven't infected PredicateInfo with this kind of special case handling yet, and I don't think this pattern is good motivation to start doing it.

ken-matsui committed:
[LVI] Infer ranges from mul nuw square conditions

A non-poison comparison involving `mul nuw X, X` implies that the
multiplication does not overflow.  This bounds X by:

  X <= floor(sqrt(2^bitwidth(X) - 1)) (e.g., i16: X <= 255)

An unsigned constant comparison can tighten the bound, e.g.,
`X * X <= 120` implies `X <= 10`.

dtcxzyw reviewed (commented):
I'd suggest implementing a new transition function like `ConstantRange::sqrt`, which interprets the input as unsigned integers and calculates the square root. You can reuse the existing `APInt::sqrt` utility.

dtcxzyw reviewed (approved):
LG

nikic reviewed (approved):
LGTM
